File: csharp/src/impl/PersistentSet.cs
namespace Volante.Impl
{
    using System;
    using System.Collections.Generic;
    using System.Collections;
    using Volante;

    class PersistentSet<T> : Btree<T, T>, Volante.ISet<T> where T : class,IPersistent
    {
        public PersistentSet()
        {
            type = ClassDescriptor.FieldType.tpObject;
            unique = true;
        }

        public override bool Contains(T o)
        {
            if (!o.IsPersistent())
                return false;

            Key key = new Key(o);
            IEnumerator<T> e = GetEnumerator(key, key, IterationOrder.AscentOrder);
            return e.MoveNext();
        }

        public override void Add(T o)
        {
            if (!o.IsPersistent())
                ((DatabaseImpl)Database).MakePersistent(o);
            base.Put(new Key(o), o);
        }

        public bool AddAll(ICollection<T> c)
        {
            bool modified = false;
            foreach (T o in c)
            {
                if (!o.IsPersistent())
                    ((DatabaseImpl)Database).MakePersistent(o);
                modified |= base.Put(new Key(o), o);
            }
            return modified;
        }

        public override bool Remove(T o)
        {
            try
            {
                Remove(new Key(o), o);
            }
            catch (DatabaseException x)
            {
                if (x.Code == DatabaseException.ErrorCode.KEY_NOT_FOUND)
                    return false;
                throw;
            }
            return true;
        }

        public bool ContainsAll(ICollection<T> c)
        {
            foreach (T o in c)
            {
                if (!Contains(o))
                    return false;
            }
            return true;
        }

        public bool RemoveAll(ICollection<T> c)
        {
            bool modified = false;
            foreach (T o in c)
            {
                modified |= Remove(o);
            }
            return modified;
        }

        public override bool Equals(object o)
        {
            if (o == this)
                return true;

            Volante.ISet<T> s = o as Volante.ISet<T>;
            if (s == null)
                return false;

            if (Count != s.Count)
                return false;

            return ContainsAll(s);
        }

        public override int GetHashCode()
        {
            int h = 0;
            foreach (IPersistent o in this)
            {
                h += o.Oid;
            }
            return h;
        }
    }
}

File: java/src/org/nachodb/impl/PersistentSet.java
package org.nachodb.impl;
import  org.nachodb.*;
import  java.util.*;

class PersistentSet extends Btree implements IPersistentSet { 
    PersistentSet() { 
        type = ClassDescriptor.tpObject;
        unique = true;
    }

    public boolean isEmpty() { 
        return nElems == 0;
    }

    public boolean contains(Object o) {
        if (o instanceof IPersistent) { 
            Key key = new Key((IPersistent)o);
            Iterator i = iterator(key, key, ASCENT_ORDER);
            return i.hasNext();
        }
        return false;
    }
    
    public Object[] toArray() { 
        return toPersistentArray();
    }

    public Object[] toArray(Object a[]) { 
        return toPersistentArray((IPersistent[])a);
    }

    public boolean add(Object o) { 
        IPersistent obj = (IPersistent)o;
        if (!obj.isPersistent()) { 
            ((StorageImpl)getStorage()).makePersistent(obj);
        }
        return put(new Key(obj), obj);
    }

    public boolean remove(Object o) { 
        IPersistent obj = (IPersistent)o;
        try { 
            remove(new Key(obj), obj);
        } catch (StorageError x) { 
            if (x.getErrorCode() == StorageError.KEY_NOT_FOUND) { 
                return false;
            }
            throw x;
        }
        return true;
    }
    
    public boolean containsAll(Collection c) { 
        Iterator i = c.iterator();
        while (i.hasNext()) { 
            if (!contains(i.next()))
                return false;
        }
        return true;
    }

    
    public boolean addAll(Collection c) {
        boolean modified = false;
        Iterator i = c.iterator();
        while (i.hasNext()) {
            modified |= add(i.next());
        }
        return modified;
    }

    public boolean retainAll(Collection c) {
        ArrayList toBeRemoved = new ArrayList();
        Iterator i = iterator();
        while (i.hasNext()) {
            Object o = i.next();
            if (!c.contains(o)) {
                toBeRemoved.add(o);
            }
        }
        int n = toBeRemoved.size();
        for (int j = 0; j < n; j++) { 
            remove(toBeRemoved.get(j));
        }
        return n != 0;         
    }

    public boolean removeAll(Collection c) {
        boolean modified = false;
        Iterator i = c.iterator();
        while (i.hasNext()) {
            modified |= remove(i.next());
        }
        return modified;
    }

    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (!(o instanceof Set)) {
            return false;
        }
        Collection c = (Collection) o;
        if (c.size() != size()) {
            return false;
        }
        return containsAll(c);
    }

    public int hashCode() {
        int h = 0;
        Iterator i = iterator();
        while (i.hasNext()) {
            h += ((IPersistent)i.next()).getOid();
        }
        return h;
    }
}
